core-bundle/contao/widgets/RowWizard.php

```diff
@@ -0,0 +1,278 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao;
+
+/**
+ * Provides methods to handle multiple rows, with each row containing multiple widgets.
+ */
+class RowWizard extends Widget
+{
+	protected $blnSubmitInput = true;
+
+	protected $strTemplate = 'be_widget_lbl';
+
+	protected array $arrFields = array();
+
+	private int|null $min = null;
+
+	private int|null $max = null;
+
+	private bool $sortable = true;
+
+	private array $actions = array('copy', 'delete');
+
+	private array $widgets = array();
+
+	public function __construct($arrAttributes = null)
+	{
+		parent::__construct($arrAttributes);
+
+		$this->preserveTags = true;
+		$this->decodeEntities = true;
+
+		System::loadLanguageFile('default');
+	}
+
+	/**
+	 * Add specific attributes.
+	 */
+	public function __set($strKey, $varValue): void
+	{
+		switch ($strKey)
+		{
+			case 'mandatory':
+				if ($varValue)
+				{
+					$this->arrAttributes['required'] = 'required';
+				}
+				else
+				{
+					unset($this->arrAttributes['required']);
+				}
+				parent::__set($strKey, $varValue);
+				break;
+
+			case 'fields':
+				$this->arrFields = $varValue;
+				break;
+
+			case 'min':
+				$this->min = $varValue ?? null;
+				break;
+
+			case 'max':
+				$this->max = $varValue ?? null;
+				break;
+
+			case 'sortable':
+				$this->sortable = (bool) $varValue;
+				break;
+
+			case 'actions':
+				if (\is_array($varValue))
+				{
+					$this->actions = array_intersect(array('copy', 'delete', 'enable'), $varValue);
+				}
+				break;
+
+			case 'style':
+				$this->style = $varValue ?? null;
+				break;
+
+			default:
+				parent::__set($strKey, $varValue);
+				break;
+		}
+	}
+
+	public function __get($strKey)
+	{
+		return match ($strKey)
+		{
+			'fields' => $this->arrFields,
+			'min' => $this->min,
+			'max' => $this->max,
+			'sortable' => $this->sortable,
+			'actions' => $this->actions,
+			default => parent::__get($strKey),
+		};
+	}
+
+	public function validate(): void
+	{
+		$varValue = array();
+		$varPost = $this->getPost($this->strName);
+
+		for ($i = 0, $c = \count($varPost); $i < $c; ++$i)
+		{
+			foreach ($this->arrFields as $key => $options)
+			{
+				$widget = $this->prepareWidget($key, $this->varValue[$i][$key] ?? null, $options, $i);
+
+				if (null === $widget)
+				{
+					continue;
+				}
+
+				$widget->validate();
+
+				if ($widget->hasErrors())
+				{
+					$this->objDca->noReload = true;
+				}
+				else
+				{
+					$varValue[$i][$key] = $widget->value;
+				}
+			}
+		}
+
+		if ($this->hasErrors())
+		{
+			$this->class = 'error';
+		}
```

Comment:
When does this trigger? 
Looks like we never call `$this->addError()`, so `hasErrors()` would always be false, wouldn’t it?

Reply:
I'll implement rendering an error when any of the fields will throw an error